### ISSUE TITLE
Problems with Borland Turbo Pascal and dynamic recompiler

### ISSUE THREAD

**Issue by Teppic1** [86Box/86Box#4226]
### What happened?

With any Pentium machine if you run Turbo Pascal (tried 6 and 7) in a DOS session in Windows 3.x it becomes unusable. It freezes every few seconds for up to 15-20 seconds. This affects both the real mode turbo.exe and the protected mode tpx.exe.

On 486s with dynamic recompiler turned off it runs without any problems at all. When I turned the recompiler on on the 486, the problem comes back. On the 386 there's also no problem (obviously this cannot run the dynamic compiler).

I haven't tried it in Windows 95/98 DOS sessions.

I'm 90% sure this problem was on pcem years ago too.

### Configuration file

```ini
[General]
vid_renderer = qt_software
force_43 = 1

[Machine]
machine = tx97
cpu_family = pentium_p54c
cpu_speed = 75000000
cpu_multi = 1.5
cpu_use_dynarec = 1
fpu_softfloat = 1
time_sync = local
fpu_type = internal
mem_size = 16384

[Video]
gfxcard = tvga8900d

[Input devices]
mouse_type = ps2

[Sound]
fm_driver = nuked

[Network]
net_01_link = 0
net_02_link = 0
net_03_link = 0
net_04_link = 0

[Storage controllers]
hdc = internal
cassette_mode = load

[Hard disks]
hdd_01_parameters = 17, 15, 900, 0, ide
hdd_01_ide_channel = 0:0
hdd_01_fn = images/msdos6.vhd
hdd_01_vhd_blocksize = 1024
hdd_01_speed = ramdisk

[Floppy and CD-ROM drives]
fdd_01_type = 35_2hd
fdd_01_turbo = 1
```

### Operating system

Windows 10

### CPU

i7 7500

### 86Box version

5582

### Build architecture

Windows - x64 (64-bit)

### Build type

- [ ] New recompiler
- [ ] Debug build

### Download source

Official website (Jenkins, GitHub)

### Additional context

Added vhd image with DOS/Win 3.11/TP7 for above config. The TP folder can be copied elsewhere for testing as nothing is installed/updated outside of it.
[msdos6.zip](https://github.com/86Box/86Box/files/14460547/msdos6.zip)

**Comment by Fenix770** [comment]
I think there is a patch for Pascal 7.0 required for Pentium 200 Mhz and faster machine. Since 86box does not emulate same RAM speeds like on the real hardware it may affect slower 486 mchines as well.

**Comment by Teppic1** [comment]
That patch is for programs compiled in Pascal that have runtime errors, so isn't related to this.

**Comment by OBattler** [comment]
Does the emulator go below 100% during the freezes?

**Comment by Teppic1** [comment]
No, it stays at 100%. I tried it on my Ryzen 7800X3D system as well with the same freezes and 100% speed.

**Comment by Sunspark-007** [comment]
@Teppic1 Two things to try, 1) Install FreeDOS 1.3 and copy your installation directory of TP 6 or 7 over to it and see if it works. If you do this, just mount the live cd in 86Box and do a "plain dos" installation to a new drive image. 2) Try using TP 5.5, see if it makes any difference with your freezes.. http://altd.embarcadero.com/download/museum/tp55.zip

I didn't try running any programs with it, but I did manage to install 5.5 which was a problematic.. everything is so fussy.. after all these years I am remembering why MS really wanted to get away from DOS. So, to install TP 5.5 with expanded memory on, running the unpack.com command causes the machine to reboot. With expanded memory off, no reboot.. ok let's try the install now.. it doesn't install! It just freezes.. Switch into FreeDOS 1.3, the install works fine there.. maybe it has to do with conventional RAM? there's a lot more free conventional memory available in FreeDOS. The installer will not work properly in MS-DOS 6.22 with 615K of conventional free. Easy enough to copy the directory over between drives though.

**Comment by Teppic1** [comment]
It's almost certainly not going to be a DOS problem, this was one of the most used DOS applications. Also the fact that it runs without any issue with the dynamic recompiler turned off, and runs perfectly fine in VirtualBox etc. means it's an emulation bug.

**Comment by OBattler** [comment]
It may be doing something the recompiler doesn't like, such as input looping on some I/O port. There's nothing I can do that. The interpreter would not have this problem as timers are processed after every instruction rather than after every block of instructions, and VirtualBox uses hardware virtualization so it's a completely different beast.

**Comment by OBattler** [comment]
You uploaded the hard disk image and the config but not the NVR folder, can you upload that please?

Also, which recompiler are you using? Old or new? And did you try the other one?

**Comment by OBattler** [comment]
This is the default idle loop not liking the recompiler. It's a CANTFIX. But I just found out that adding DEVICE=C:\DOS\POWER.EXE (or using any other DOS idle driver) makes it work.

**Comment by Sunspark-007** [comment]
Interesting.. trying power.exe in a dos-only environment increases the CPU load on my system, possibly why the line was REM'ed out in my config.sys

Trying out MS VPC's IDLE.COM drops cpu load from 33% to 19% on this system. FreeDOS's FDAPM is the same.

Interestingly, emulated PS/2 mouse movements are smoother with IDLE on than without in a game's intro screen.
